File: mergesort.py
list = [4, 3, 5, 7, 6, 2, 1, 9, 8, 10]

# Function to merge two sorted halves of a list
def merge(list, left, right):
    sorted = []  # Temporary list to store sorted elements
    i = left  # Starting index of the left half
    mid = (left + right) // 2  # Midpoint index to divide the list into halves
    j = mid + 1  # Starting index of the right half

    # Merge elements from both halves into sorted list
    while i < mid + 1 and j < right + 1:
        if list[i] <= list[j]:
            sorted.append(list[i])
           
            i = i + 1
        else:
            sorted.append(list[j])
            j = j + 1
    #for x in range(len(sorted)):
        #print("temporary array: ",sorted[x])
    #print()
    # Add remaining elements from the right half, if any
    while j < right + 1:
        sorted.append(list[j])
        j = j + 1

    # Add remaining elements from the left half, if any
    while i < mid + 1:
        sorted.append(list[i])
        i = i + 1

    # Copy sorted elements back to the original list
    for x in range(len(sorted)):
        list[left + x] = sorted[x]

# Function to recursively divide and merge the list
def mergesort(list, left, right):
    mid = (left + right) // 2  # Calculate the midpoint index

    # Base case: if there is only one element in the sublist
    if right==left:
        print("base case:",left)
        print()
        return
    else:
        print("start of else:", "left:",left,"right:", right)
        print()
        # Recursive calls to divide the list into two halves
        mergesort(list, left, mid)
        #stack is created, so the mergesort happens inside itself, making the left and right smaller and smaller until the basecase is reached
        mergesort(list, mid + 1, right)
        # Merge the sorted halves
        merge(list, left, right)
        print("end of the else:",left, right)
# Call mergesort function to sort the list
mergesort(list, 0, len(list) - 1)

# Print the sorted list
for x in range(len(list)):
    print(list[x])

File: Mergesort.java
package Sorting;
public class Mergesort {
    /**
     * Merge Sort algorithm:
     * - Divides the array into two halves until each subarray contains one element (or empty).
     * - Recursively merges the sorted subarrays to produce larger sorted subarrays until the entire array is sorted.
     */
    public static void mergeSort(int[]array){
        if(array.length<=1){
            return;
        }
        int length = array.length;
        int mid = length/2;

        //creation of the temporary arrays
        int [] left=new int[mid];
        int [] right = new int[length-mid];
        for(int i = 0; i<mid; i++){
            left[i]=array[i];
        }
        for(int e = mid; e<length; e++){
            right[e-mid]=array[e];
        }
        mergeSort(left); //doing mergeSort on the left
        mergeSort(right); //doing mergeSort on the right
        merge(array, left, right);
    }
    private static void merge(int [] array, int[]left, int[]right){
        int leftlength=left.length;
        int rightlength=right.length;
        //we need 3 index because the array lengths will change, so we need to look at different parts of the arrays 
        int xleft=0;
        int xright=0;
        int x=0;
        while(xleft<leftlength && xright<rightlength){
            if(left[xleft]>right[xright]){
                array[x]=right[xright];
                x+=1;
                xright+=1;
            }
            else{
            array[x]=left[xleft];
            x+=1;
            xleft+=1;   
            }
        }
        //after one of the two split arrays are fully added into the big array, these while loops make sure the remaining array is fully added into it, *note* the remaining array is already sorted
        while(xleft<leftlength){
            array[x]=left[xleft];
            x+=1;
            xleft+=1;
        }
        while(xright<rightlength){
            array[x]=right[xright];
            x+=1;
            xright+=1;
        }
    }
    public static void main(String[]args){
        int[]Array={4,3,9,8,5,10,2,7};
        mergeSort(Array);
        for(int x = 0; x<Array.length; x++){
            System.out.println(Array[x]);
        }
    }
}
